Library/Homebrew/cmd/test-bot.rb

```diff
@@ -944,11 +953,22 @@ def test_bot
 
           if step.has_output?
             # Remove invalid XML CData characters from step output.
-            output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            if ruby_has_encoding?
+              # This is the regex for valid XML chars, but only works in Ruby 2.0+
+              # /[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
+              # For 1.9 compatibility, use the inverse of that, which stays under \u10000
+              invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+              output = step.output.gsub(invalid_xml_pat, "\uFFFD")
+            else
+              output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, output.length - BYTES_IN_1_MEGABYTE].max
+              if slice_start > 0
+                output = "truncated output to ~1MB:\n" \
+                  + output.slice(slice_start, output.length - slice_start)
+              end
```

Comment:
Thanks. I was aware of it (mentioned it in the original issue post), just didn't think it was worth taking care of. But the "truncate in BINARY and transcode" fix you suggested is pretty straightforward, so I'm going to go ahead and try it here.